<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1" />
        <meta name="description" content="" />
        <meta name="author" content="Gregor Sturm" />
        <title>LuCA - single-cell lung cancer atlas</title>

        <link rel="canonical" href="https://luca.icbi.at" />

        <link href="./css/bootstrap.min.css" rel="stylesheet" />

        <style>
            .bd-placeholder-img {
                font-size: 1.125rem;
                text-anchor: middle;
                -webkit-user-select: none;
                -moz-user-select: none;
                user-select: none;
            }

            @media (min-width: 768px) {
                .bd-placeholder-img-lg {
                    font-size: 3.5rem;
                }
            }
        </style>

        <link href="custom.css" rel="stylesheet" />
    </head>
    <body>
        <div class="col-lg-4 mx-auto p-3 py-md-5">
            <header class="d-flex align-items-center pb-3 mb-5 border-bottom">
                <a
                    href="/"
                    class="d-flex align-items-center text-dark text-decoration-none"
                >
                    <span class="fs-4">LuCA</span>
                </a>
            </header>

            <main>
                <h1>Single-cell <b>Lu</b>ng <b>C</b>ancer <b>A</b>tlas</h1>
                <p class="fs-5 col-md-12"></p>
                <figure>
                    <blockquote class="blockquote">
                        High-resolution single-cell atlas reveals diversity and
                        plasticity of tissue-resident neutrophils in non-small
                        cell lung cancer
                    </blockquote>
                    <figcaption class="blockquote-footer">
                        Salcher, Sturm, Horvath et al.
                        <cite
                            >BioRxiv,
                            <a href="https://doi.org/10.1101/2022.05.09.491204"
                                >doi: 10.1101/2022.05.09.491204</a
                            ></cite
                        >
                    </figcaption>
                </figure>

                <p>
                    The core atlas was compiled from 21 datasets comprising 505
                    samples from 298 patients. 38 different cell-types were
                    annotated at different levels of granularity.
                </p>

                <p>
                    By using the transfer learning method
                    <a href="https://scarches.readthedocs.io/en/latest/"
                        >scArches</a
                    >
                    another 8 datasets were projected onto the core atlas,
                    resulting in an extended atlas with
                    <b>1.3 million cells, 556 samples and 318 patients.</b>
                </p>

                <div class="d-flex justify-content-center">
                    <img
                        class="full-width"
                        src="img/core_atlas_umap_1000px.png"
                        alt="Atlas overview UMAP"
                    />
                </div>

                <h2>CZI cell-x-gene portal</h2>
                <p>
                    Both core and extended atlas are available on the CZI
                    cell-x-gene. There, you can
                    <b>explore the datasets interactively</b> in the browser and
                    <b>download them</b> as <code>h5ad</code> (<a
                        href="https://scverse.org"
                        >scverse</a
                    >) or <code>rds</code> (Seurat) files with standardized
                    metadata for further analysis.
                </p>

                <div class="row g-5">
                    <div class="col-md-12">
                        <div class="float-end">
                            <a
                                href="https://cellxgene.cziscience.com/collections/edb893ee-4066-4128-9aec-5eb2b03f8287"
                                class="btn btn-outline-primary btn-lg px-4"
                                ><svg
                                    xmlns="http://www.w3.org/2000/svg"
                                    width="16"
                                    height="16"
                                    fill="currentColor"
                                    class="bi bi-caret-right"
                                    viewBox="0 0 16 16"
                                >
                                    <path
                                        d="M6 12.796V3.204L11.481 8 6 12.796zm.659.753 5.48-4.796a1 1 0 0 0 0-1.506L6.66 2.451C6.011 1.885 5 2.345 5 3.204v9.592a1 1 0 0 0 1.659.753z"
                                    />
                                </svg>
                                Open CZI portal</a
                            >
                        </div>
                    </div>
                </div>

                <h2>scArches model</h2>
                <p>
                    For the core atlas, we also provide a
                    <a href="https://scarches.readthedocs.io/en/latest/"
                        >scArches</a
                    >
                    model that allows to project new datasets onto the atlas and
                    automatically annotate cell-types.
                </p>

                <div class="row g-5">
                    <div class="col-md-12">
                        <div class="float-end">
                            <a
                                href="https://zenodo.org/record/6997383/files/core_atlas_scanvi_model.tar.gz?download=1"
                                class="btn btn-outline-primary btn-lg px-4"
                                ><svg
                                    xmlns="http://www.w3.org/2000/svg"
                                    width="16"
                                    height="16"
                                    fill="currentColor"
                                    class="bi bi-cloud-download"
                                    viewBox="0 0 16 16"
                                >
                                    <path
                                        d="M4.406 1.342A5.53 5.53 0 0 1 8 0c2.69 0 4.923 2 5.166 4.579C14.758 4.804 16 6.137 16 7.773 16 9.569 14.502 11 12.687 11H10a.5.5 0 0 1 0-1h2.688C13.979 10 15 8.988 15 7.773c0-1.216-1.02-2.228-2.313-2.228h-.5v-.5C12.188 2.825 10.328 1 8 1a4.53 4.53 0 0 0-2.941 1.1c-.757.652-1.153 1.438-1.153 2.055v.448l-.445.049C2.064 4.805 1 5.952 1 7.318 1 8.785 2.23 10 3.781 10H6a.5.5 0 0 1 0 1H3.781C1.708 11 0 9.366 0 7.318c0-1.763 1.266-3.223 2.942-3.593.143-.863.698-1.723 1.464-2.383z"
                                    />
                                    <path
                                        d="M7.646 15.854a.5.5 0 0 0 .708 0l3-3a.5.5 0 0 0-.708-.708L8.5 14.293V5.5a.5.5 0 0 0-1 0v8.793l-2.146-2.147a.5.5 0 0 0-.708.708l3 3z"
                                    />
                                </svg>
                                core atlas (scArches)</a
                            >
                        </div>
                    </div>
                </div>

                <h2>Get input data and intermediate results</h2>
                <p>
                    In addition to the <code>h5ad</code> objects mentioned
                    above, we provide preprocessed input data, intermediate
                    results and final results on zenodo. The results were
                    derived using the source code provided on GitHub and contain
                    a rendered version of the Jupyter/Rmarkdown notebooks that
                    were used to generate the figures in the manuscript.
                </p>
                <div class="row g-5">
                    <div class="col-md-12">
                        <a
                            href="https://doi.org/10.5281/zenodo.6411867"
                            class="btn btn-outline-primary btn-lg px-4 float-end"
                            ><svg
                                xmlns="http://www.w3.org/2000/svg"
                                width="16"
                                height="16"
                                fill="currentColor"
                                class="bi bi-cloud-download"
                                viewBox="0 0 16 16"
                            >
                                <path
                                    d="M4.406 1.342A5.53 5.53 0 0 1 8 0c2.69 0 4.923 2 5.166 4.579C14.758 4.804 16 6.137 16 7.773 16 9.569 14.502 11 12.687 11H10a.5.5 0 0 1 0-1h2.688C13.979 10 15 8.988 15 7.773c0-1.216-1.02-2.228-2.313-2.228h-.5v-.5C12.188 2.825 10.328 1 8 1a4.53 4.53 0 0 0-2.941 1.1c-.757.652-1.153 1.438-1.153 2.055v.448l-.445.049C2.064 4.805 1 5.952 1 7.318 1 8.785 2.23 10 3.781 10H6a.5.5 0 0 1 0 1H3.781C1.708 11 0 9.366 0 7.318c0-1.763 1.266-3.223 2.942-3.593.143-.863.698-1.723 1.464-2.383z"
                                />
                                <path
                                    d="M7.646 15.854a.5.5 0 0 0 .708 0l3-3a.5.5 0 0 0-.708-.708L8.5 14.293V5.5a.5.5 0 0 0-1 0v8.793l-2.146-2.147a.5.5 0 0 0-.708.708l3 3z"
                                />
                            </svg>
                            zenodo</a
                        >
                    </div>
                </div>

                <h2>Browse code</h2>
                <p>
                    All code needed to reproduce the study is wrapped into a
                    nextflow workflow and publicy available from GitHub.
                </p>

                <div class="row g-5">
                    <div class="col-md-12">
                        <a
                            href="https://github.com/icbi-lab/luca"
                            class="btn btn-outline-primary btn-lg px-4 float-end"
                            ><svg
                                xmlns="http://www.w3.org/2000/svg"
                                width="16"
                                height="16"
                                fill="currentColor"
                                class="bi bi-github"
                                viewBox="0 0 16 16"
                            >
                                <path
                                    d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.012 8.012 0 0 0 16 8c0-4.42-3.58-8-8-8z"
                                />
                            </svg>
                            icbi-lab/luca</a
                        >
                    </div>
                </div>

                <h2>Contact</h2>
                <p>
                    For reproducibility issues or any other requests regarding
                    single-cell data analysis, please use the
                    <a href="https://github.com/icbi-lab/luca/issues"
                        >issue tracker</a
                    >. For anything else, you can reach out to the corresponding
                    author(s) as indicated in the manuscript.
                </p>
            </main>

            <footer class="pt-5 my-5 text-muted border-top">
                &copy; 2022
                <a href="https://icbi.i-med.ac.at/"
                    >Institute of Bioinformatics, Medical University of
                    Innsbruck</a
                >
                &middot;
                <a href="https://icbi.i-med.ac.at/imprint.html">Impressum</a>
            </footer>
        </div>
    </body>
</html>
